### PR TITLE
CLDR-15653 LocaleCompletion: cache, concurrency, metrics

### PR DESCRIPTION
**Pull request by srl295**
- update Summary (dashboard) for metric names
- add metrics to LocaleCompletion
- add a Cache to LocaleCompletion, that gets invalidated if any votes change

CLDR-15653

- [ ] This PR completes the ticket.

### CODE REVIEW

**Comment by srl295** [comment]
may still be a sql issue, but this mitigates any load/concurrency issue

File: tools/cldr-apps/src/main/java/org/unicode/cldr/web/api/LocaleCompletion.java [hunk 1]
```diff
@@ -83,39 +95,68 @@ public class LocaleCompletion {
...
 
-    static final class LocaleCompletionHelper {
+    static final class LocaleCompletionHelper implements Listener {
         PathHeader.Factory phf = null;
-
+        LoadingCache<CLDRLocale, LocaleCompletionResponse> cache;
         LocaleCompletionHelper() {
             phf = PathHeader.getFactory(CLDRConfig.getInstance().getEnglish());
+            cache =  CacheBuilder.newBuilder().maximumSize(500)
+            // We would want the expiry time greater if this were used by users, not just TC.
```
Comment (btangmu) [inline tools/cldr-apps/src/main/java/org/unicode/cldr/web/api/LocaleCompletion.java:131]: it's not clear to me why this is "just TC" -- is this for the progress meter, also used by regular vetters?

File: tools/cldr-apps/src/main/java/org/unicode/cldr/web/api/LocaleCompletion.java [hunk 2]
```diff
@@ -2,15 +2,20 @@
...
+import javax.enterprise.context.ApplicationScoped;
 import javax.ws.rs.GET;
 import javax.ws.rs.Path;
 import javax.ws.rs.PathParam;
 import javax.ws.rs.Produces;
 import javax.ws.rs.core.MediaType;
 import javax.ws.rs.core.Response;
 
+import org.eclipse.microprofile.metrics.annotation.Counted;
+import org.eclipse.microprofile.metrics.annotation.Timed;
```
Comment (btangmu) [inline tools/cldr-apps/src/main/java/org/unicode/cldr/web/api/LocaleCompletion.java:18]: is this our first use of microprofile.metrics?

File: tools/cldr-apps/src/main/java/org/unicode/cldr/web/api/LocaleCompletion.java [hunk 3]
```diff
@@ -83,39 +95,68 @@ public class LocaleCompletion {
...
+            });
         }
-
         static LocaleCompletionHelper INSTANCE = new LocaleCompletionHelper();
+
+        @Override
+        public void valueChanged(String xpath, XMLSource source) {
+            // invalidate the named entry
+            // this is registered whenever a calculation happens
+            System.err.println("LCH Invalidating " + source.getLocaleID());
```
Comment (btangmu) [inline tools/cldr-apps/src/main/java/org/unicode/cldr/web/api/LocaleCompletion.java:147]: System.err.println -- probably should be only when debugging? and "info" not "err"?

File: tools/cldr-apps/src/main/java/org/unicode/cldr/web/api/LocaleCompletion.java [hunk 4]
```diff
@@ -83,39 +95,68 @@ public class LocaleCompletion {
...
+            return STError.surveyNotQuiteReady();
+        }
         CLDRLocale cldrLocale = CLDRLocale.getInstance(localeId);
-        return getLocaleCompletion(cldrLocale);
+        return Response.ok(getLocaleCompletion(cldrLocale)).build();
+    }
+
+    static LocaleCompletionResponse getLocaleCompletion(CLDRLocale cldrLocale) throws ExecutionException {
+        return LocaleCompletionHelper.INSTANCE.cache.get(cldrLocale);
+        // return handleGetLocaleCompletion(cldrLocale);
```
Comment (btangmu) [inline tools/cldr-apps/src/main/java/org/unicode/cldr/web/api/LocaleCompletion.java:111]: don't need commented-out code?

File: tools/cldr-apps/src/main/java/org/unicode/cldr/web/api/LocaleCompletion.java [hunk 5]
```diff
@@ -83,39 +95,68 @@ public class LocaleCompletion {
...
+    static final class LocaleCompletionHelper implements Listener {
         PathHeader.Factory phf = null;
-
+        LoadingCache<CLDRLocale, LocaleCompletionResponse> cache;
         LocaleCompletionHelper() {
             phf = PathHeader.getFactory(CLDRConfig.getInstance().getEnglish());
+            cache =  CacheBuilder.newBuilder().maximumSize(500)
+            // We would want the expiry time greater if this were used by users, not just TC.
+            .expireAfterWrite(15, TimeUnit.MINUTES)
+            // .concurrencyLevel()
```
Comment (btangmu) [inline tools/cldr-apps/src/main/java/org/unicode/cldr/web/api/LocaleCompletion.java:133]: don't need commented-out code

File: tools/cldr-apps/src/main/java/org/unicode/cldr/web/api/LocaleCompletion.java [hunk 6]
```diff
@@ -83,39 +95,68 @@ public class LocaleCompletion {
...
+            // this is registered whenever a calculation happens
+            System.err.println("LCH Invalidating " + source.getLocaleID());
+            cache.invalidate(CLDRLocale.getInstance(source.getLocaleID()));
+        }
     }
 
-    static LocaleCompletionResponse getLocaleCompletion(final CLDRLocale cldrLocale, final STFactory stFactory) {
+    /**
+     *
+     * This function computes the actual locale completion given a Locale abd STFactory
```
Comment (btangmu) [inline tools/cldr-apps/src/main/java/org/unicode/cldr/web/api/LocaleCompletion.java:154]: typo "abd"

File: tools/cldr-apps/src/main/java/org/unicode/cldr/web/api/LocaleCompletion.java (hunk 2 above)
Comment (btangmu) [reply tools/cldr-apps/src/main/java/org/unicode/cldr/web/api/LocaleCompletion.java:18]: no i see we already had it in Summary.java -- awesome!

**Review by btangmu — APPROVED** [review]
suggested minor improvements could be follow-up pr

File: tools/cldr-apps/src/main/java/org/unicode/cldr/web/api/LocaleCompletion.java (hunk 1 above)
Comment (srl295) [reply tools/cldr-apps/src/main/java/org/unicode/cldr/web/api/LocaleCompletion.java:131]: Copy and paste error from an unrelated cache! Ignore the comment

**Comment by btangmu** [comment]
> may still be a sql issue, but this mitigates any load/concurrency issue

"sql issue" reminded me of something I saw in the log:
```
openliberty@st:/var/log/openliberty/cldr$ egrep vote_value messages_22.05.17_13.47.08.0.log
[5/17/22, 7:26:18:640 UTC] 00000077 SystemOut                                                    O SQL: select cldr_vote_value_42.locale,cldr_vote_value_42.xpath,cldr_users.org, cldr_vote_value_42.value, cldr_vote_value_42.last_mod  from cldr_vote_value_42,cldr_users  where  cldr_users.id=cldr_vote_value_42.submitter  order by cldr_vote_value_42.last_mod desc 
```
where did that come from? it happened at 7:26, right before LocaleCompletion started starting without stopping at 7:29!

**Comment by srl295** [comment]
I tested in the Ui that this invalidates correctly by the way. So if you make a changed vote it will update the completion meter right away.

**Pull request merged**